### PR TITLE
Add challenge: DPO Sequence Loss

### PR DESCRIPTION
**Pull request by basilwong**
## Summary

  Adds **easy challenge 108: DPO Sequence Loss**.

The challenge implements Direct Preference Optimization (DPO) over chosen and rejected responses. It reduces token-level log probabilities to sequence-level policy and reference margins, applies the DPO preference objective, and returns the mean scalar loss across the batch.

  ## Scope

  - Inputs:
    - chosen and rejected policy log probabilities: `[B, S]`
    - chosen and rejected reference-model log probabilities: `[B, S]`
  - Each chosen/rejected pair represents responses to the same prompt and must use the same tokenization and token-position convention.
  - Token-level log probabilities are reduced to sequence-level values before computing the preference margin.
  - The reference-model inputs are fixed precomputed values; gradients must not flow through them.
  - Padding and token masks are outside the scope of this challenge: all `S` positions participate in each sequence reduction.
  - The challenge is classified as **easy** because it is a direct, fully specified implementation of the DPO loss.

  ## Documentation and validation

  - Documents tensor shapes, sequence reduction, the role of the reference-model margin, and numerical expectations.
  - Adds a test with a nonzero reference-model margin to ensure the implementation does not accidentally omit the reference-policy correction.
  - Keeps the implementation focused on the core DPO objective without introducing a model forward pass, data pipeline, or reward-model training.

  Modal runtime validation notebook: [DPO Sequence Loss notebook](https://modal.com/notebooks/basilwong/main/nb-DpBYTyYi47iqxqkvI1wxFl)

  ## Reference

  [Direct Preference Optimization: Your Language Model is Secretly a Reward Model — Rafailov et al. (2023)](https://arxiv.org/abs/2305.18290)

### CODE REVIEW

**Comment by basilwong** [comment]
## Related Challenges

  - [Sigmoid](https://github.com/AlphaGPU/leetgpu-challenges/blob/main/challenges/easy/68_sigmoid/challenge.html)

    DPO uses a log-sigmoid preference objective, so it shares the sigmoid nonlinearity with this challenge. The Sigmoid challenge is a simple element-wise activation, whereas DPO constructs a sequence-level chosen-versus-rejected preference margin, corrects it using fixed reference-model log probabilities, then computes a mean scalar loss.

  - [Logistic Regression](https://github.com/AlphaGPU/leetgpu-challenges/blob/main/challenges/medium/34_logistic_regression/challenge.html)

    Both DPO and logistic regression use a logistic likelihood, but their tasks differ substantially. Logistic Regression optimizes model coefficients from feature vectors and
    binary labels; DPO evaluates a fixed pairwise preference objective from policy and reference log probabilities, without a feature matrix, iterative fitting loop, or learned
    classifier parameters.

  - [Categorical Cross-Entropy Loss](https://github.com/AlphaGPU/leetgpu-challenges/blob/main/challenges/medium/25_categorical_cross_entropy_loss/challenge.html)

    Both challenges produce a scalar likelihood-based loss, but categorical cross-entropy scores a target class from a multiclass logit vector. DPO instead compares two full response sequences for the same prompt and uses the reference-model margin to define the preference score.

File: challenges/easy/108_dpo_sequence_loss/challenge.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+from typing import Any, Dict, List
+
+import torch
+import torch.nn.functional as F
+from core.challenge_base import ChallengeBase, OutTensor, RandnTensor
+
+
+class Challenge(ChallengeBase):
+    name = "DPO Sequence Loss"
+    atol = 0.001
```
Comment (shxjames) [inline challenges/easy/108_dpo_sequence_loss/challenge.py]: Same as before: need to make the tolerance a bit tighter

File: challenges/medium/108_dpo_sequence_loss/challenge.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        return self._make_test_case(
+            4,
+            chosen_logps=[0.0, 1.0, -1.0, 2.0],
+            rejected_logps=[0.0, 0.0, 0.0, 0.0],
+            chosen_ref_logps=[0.0, 0.0, 0.0, 0.0],
+            rejected_ref_logps=[0.0, 0.0, 0.0, 0.0],
+            beta=0.1,
+        )
+
+    def generate_functional_test(self) -> List[Dict[str, Any]]:
```
Comment (shxjames) [inline challenges/medium/108_dpo_sequence_loss/challenge.py:111]: Perhaps just a choice towards being more "realistic": we should consider adding more test cases where the logps have 1. more sequence length, and 2. more negative values. As probs are really low for actuall llms, their logps will also be quite small.

Comment (basilwong) [reply challenges/medium/108_dpo_sequence_loss/challenge.py:111]: Makes sense. 

Since this challenge takes pre-summed sequence log probabilities, I modelled longer responses as larger-magnitude negative log-probability values rather than adding a token-length dimension.

I added a couple deterministic cases with sequence-level log probabilities around `-400` to `-2000`, including:
  - policy and reference margins that remain close at realistic magnitudes
  - short chosen versus long rejected responses, where the large length-driven margin must cancel against the reference-policy margin

**Comment by basilwong** [comment]
Thanks for the feedback @shxjames !! Here are the following PR updates:

Challenge Description Updates:
- Updated the description’s math and examples for clearer frontend rendering.
- Added a visual explanation DPO

image

---

Added functional tests: 

  - Tightened the `beta` contract to `0 < beta <= 1`, matching the DPO formulation. The degenerate `beta = 0` case was removed.
  - Added functional coverage for:
    - `beta = 1` with a nonzero preference logit; and
    - nonzero current/reference margins that cancel exactly to `z = 0`.
  - The existing tests retain realistic large negative sequence-level log probabilities and length-driven margins, including cases that require the reference-policy correction.

---

  The updated [Modal notebook](https://modal.com/notebooks/basilwong/main/nb-DpBYTyYi47iqxqkvI1wxFl) deterministically verifies that the example implementations pass the functional tests.

**Review by shxjames — APPROVED** [review]
